V7

printercu opened the pull request:
Hi all, @RubyMoney/money-devs!

I had some ideas to implement in Money for a while. Most of them are about overall simplification of source code, with some features. I've got some free time and started codding. Can you please take a look. Ideas and critics are welcome.

Ready:
- [x] lib/money/money/\* moved to lib/money.
  Ex., `Money::Currency` is in the `money/currency`, not `money/money/currency`.
- [x] formating specific methods moved to Formatter.
  This allows to define custom formatters and change formatting rules easily.
  - default_formatting_rules & use_i18n moved to Formatter.
  - Formatter.default_rules= doesn't support nil (hash must be given)
- [x] `.default_currency` doesn't wraps lambda's result with Currency.
  Simplifies code.
- [x] Rename `#iso_code` to `#code` in Currency.
  Better support for non-ISO currencies.

TODO, to discuss:
- [x] use ruby 2 features.
  Keyword arguments.
- [x] remove currency_as_string.
  Changelog says it's for rails, but there is no such calls in rails's master.
- [x] simplify arithmetics.
  Remove ambiguous coercions, encourage using to_d.
- [x] simplify #to_s or change it to use its own formatter.
  It's really complex now, does anybody use it instead of `#format`)?
- [x] simplify default formatter.
  Leaving current formatter still available?
- [x] use whole amounts (not cents) as main value
  Providing compatibility as separate module.
- [ ] change currency stores from json to yaml.
  It's human-friendly.
- [x] make Currency's '#id', '#to_sym' return same values.
  For now id is in downcase, to_sym - upcase.
- [x] make Currency#id == Currency#code.to_sym

All test from v6 are passing. I haven't change them a lot, just fixed values for `.default_currency` & `.default_rules`.

HaleTom commented:
There was one other item I'm aware of which was waiting for a major version bump. 

`.code` actually returns a currency _symbol_, unless the symbol is not defined, in which case it returns a code string like 'USD'.

```
    def code
      symbol || iso_code
    end
```
This is counter-intuitive, but fixing it will probably break legacy code, hence waiting for a major version bump.

- [ ]  Have `.code` actually return the currency code (eg 'USD')

`.symbol` will continue to return the symbol, eg '$' for USD.

Support for non-ISO currencies:

`.iso_code` doesn't make sense for BTC and other user-defined non-ISO currencies

- [ ] Rename `.iso_code` to be just `.code`
- [ ] Keep `.iso_code` as an alias for `.code` for backwards compatibility
- [ ] Implement `is_iso?` to distinguish between the ISO and non-ISO currencies

See Issue #598 for background on all the above.

The main one which requires the version bump is `.code` actually returning a currency code, possibly breaking existing other client code.

I'd be happy to code this up if the consensus is a GO.

printercu commented:
@HaleTom yep! I've forgotten about this first. Added commit. I've also changed Currency to use `#code.downcase.to_sym` as id. So it's not repeated in currency files. However after this YEN & GHC fallbacks are not supported (JPY & GHS should be used instead).

I'll add `#is_iso?` later after changing `iso_numeric` to Int.

HaleTom commented:
Cheers! You're on fire!

antstorm commented:
@printercu great work! My suggestion is to create another branch and opening separate PRs (which can be reviewed individually) against that branch. Otherwise this is now becoming pretty much unreviewable

ollym commented:
@printercu any news? would be good to merge and release this soon

printercu commented:
I thought, nobody was interested in this pr. I can rebase and fix conflicts if you really intend to merge this one :)

printercu commented:
I've just rebased on master, and added some small commits.

I've decided to release it as `money2` gem for now, as I've been using it for a while in several projects.
Here is a note I've made to share with others: https://github.com/printercu/money/wiki

I'll be glad to receive suggestions, PRs or any other feedback.

dmitry commented:
@printercu please ask @semmons99 to be one of the maintainers: https://github.com/RubyMoney/money/issues/628 Would be really great to see all your changes in the main repo as well!

antstorm commented:
@printercu still not sure why you decided to release it separately, I though that you didn't have time or something. We were really close to merging #622.

printercu commented:
@antstorm I just want to make it easier to try to move to new version and to give early-adopter's feedback and understand overall reaction. As I said, there are just too much incompatibility, so I really don't have time now to add deprecation warnings all over the code first, and then rebase everything again, while I've already have working v7 code.

As I mentioned, there is V6Compatibility module. I think It should be released as separate gem, which can be modified to raise warnings. However I don't even know how can this be done for some deprecations (ex. changing `Money.from_amount -> Money.new`). So best for me is just to release new major version and provide robust migration guide, which can be prepared with early-adopters.

@dmitry I don't think it would be right enough to become an owner just to push own changes without community support. Main repo is improved continuously and I hope this PR or something else similar will be merged soon.

bdewater commented:
Wondering if there's any progress on this, I'm liking these changes a lot!

printercu commented:
@bdewater I'm not sure if this would be merged soon. For now I have [money2 gem](https://github.com/printercu/money), and merging money's master in sometimes. Please see https://github.com/RubyMoney/money/pull/619#issuecomment-269368526 for more info.

antstorm commented:
@printercu let's actually make this happen. I've created a new branch `v7_new`, can you please split your changes into smaller PRs and open them against that branch? I'll review everything and we'll merge it.

printercu commented:
I'll try soon

antstorm commented:
@printercu thanks, ping me if you need any help with that

AlexWayfer commented:
Should this be closed?
